commit e144c51e16dbabe601ffc2f44c42da2ec0c8f7ff
Author: Alexander Hofbauer <alex@derhofbauer.at>
Date:   Wed Jan 7 12:07:47 2015 +0100

    Make volume steps configurable

diff --git a/data/org.gnome.settings-daemon.plugins.gschema.xml.in.in b/data/org.gnome.settings-daemon.plugins.gschema.xml.in.in
index b74ed42..6239200 100644
--- a/data/org.gnome.settings-daemon.plugins.gschema.xml.in.in
+++ b/data/org.gnome.settings-daemon.plugins.gschema.xml.in.in
@@ -137,5 +137,11 @@
       <_summary>Priority to use for this plugin</_summary>
       <_description>Priority to use for this plugin in gnome-settings-daemon startup queue</_description>
     </key>
+    <key name="volume-step" type="i">
+      <default>6</default>
+      <range min="1" max="20"/>
+      <_summary>Size of volume step</_summary>
+      <_description>Size of the volume step for each consecutive press of the volume up or volume down key</_description>
+    </key>
   </schema>
 </schemalist>
diff --git a/plugins/media-keys/gsd-media-keys-manager.c b/plugins/media-keys/gsd-media-keys-manager.c
index dcd14f2..64418a1 100644
--- a/plugins/media-keys/gsd-media-keys-manager.c
+++ b/plugins/media-keys/gsd-media-keys-manager.c
@@ -100,10 +100,11 @@ static const gchar introspection_xml[] =
 #define SETTINGS_POWER_DIR "org.gnome.settings-daemon.plugins.power"
 #define SETTINGS_XSETTINGS_DIR "org.gnome.settings-daemon.plugins.xsettings"
 #define SETTINGS_TOUCHPAD_DIR "org.gnome.settings-daemon.peripherals.touchpad"
+#define SETTINGS_SOUND_DIR "org.gnome.settings-daemon.plugins.sound"
 #define TOUCHPAD_ENABLED_KEY "touchpad-enabled"
+#define VOLUME_STEP_KEY "volume-step"
 #define HIGH_CONTRAST "HighContrast"
 
-#define VOLUME_STEP 6           /* percents for one volume button press */
 #define MAX_VOLUME 65536.0
 
 #define SYSTEMD_DBUS_NAME                       "org.freedesktop.login1"
@@ -142,6 +143,7 @@ struct GsdMediaKeysManagerPrivate
         GvcMixerStream  *source;
         ca_context      *ca;
         GtkSettings     *gtksettings;
+        GSettings       *sound_settings;
 #ifdef HAVE_GUDEV
         GHashTable      *streams; /* key = X device ID, value = stream id */
         GUdevClient     *udev_client;
@@ -1173,7 +1175,7 @@ do_sound_action (GsdMediaKeysManager *manager,
 {
 	GvcMixerStream *stream;
         gboolean old_muted, new_muted;
-        guint old_vol, new_vol, norm_vol_step;
+        guint old_vol, new_vol, vol_step, norm_vol_step;
         gboolean sound_changed;
 
         /* Find the stream that corresponds to the device, if any */
@@ -1192,7 +1194,8 @@ do_sound_action (GsdMediaKeysManager *manager,
         if (stream == NULL)
                 return;
 
-        norm_vol_step = PA_VOLUME_NORM * VOLUME_STEP / 100;
+        vol_step = g_settings_get_int (manager->priv->sound_settings, VOLUME_STEP_KEY);
+        norm_vol_step = PA_VOLUME_NORM * vol_step / 100;
 
         /* FIXME: this is racy */
         new_vol = old_vol = gvc_mixer_stream_get_volume (stream);
@@ -2340,6 +2343,8 @@ start_media_keys_idle_cb (GsdMediaKeysManager *manager)
           g_hash_table_new_full (g_str_hash, g_str_equal,
                                  g_free, g_object_unref);
 
+        manager->priv->sound_settings = g_settings_new (SETTINGS_SOUND_DIR);
+
         /* for the power plugin interface code */
         manager->priv->power_settings = g_settings_new (SETTINGS_POWER_DIR);
 
@@ -2443,6 +2448,7 @@ gsd_media_keys_manager_stop (GsdMediaKeysManager *manager)
 
         g_clear_object (&priv->logind_proxy);
         g_clear_object (&priv->settings);
+        g_clear_object (&priv->sound_settings);
         g_clear_object (&priv->power_settings);
         g_clear_object (&priv->power_proxy);
         g_clear_object (&priv->power_screen_proxy);
